File: core/src/testFixtures/java/by/mrrockka/extension/CorePSQLExtension.java
package by.mrrockka.extension;

import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.extension.AfterEachCallback;
import org.junit.jupiter.api.extension.BeforeAllCallback;
import org.junit.jupiter.api.extension.ExtensionContext;

import static by.mrrockka.extension.TestPSQLContainer.*;

@Slf4j
public class CorePSQLExtension implements BeforeAllCallback, AfterEachCallback {
  @Override
  public void beforeAll(final ExtensionContext context) {
    container.start();
    datasourceProperties();
  }

  private void datasourceProperties() {
    System.setProperty("spring.datasource.url", "jdbc:tc:postgresql:%s:///%s".formatted(VERSION, DB_NAME));
    System.setProperty("spring.datasource.username", container.getUsername());
    System.setProperty("spring.datasource.password", container.getPassword());
  }

  @Override
  public void afterEach(final ExtensionContext context) {
//    todo: generate data for ITests instead of prepopulating throughout migration scripts
//    final var jdbcTemplate = SpringExtension.getApplicationContext(context).getBean(NamedParameterJdbcTemplate.class);
//    jdbcTemplate.execute("truncate bounty", PreparedStatement::execute);
//    jdbcTemplate.execute("truncate entries", PreparedStatement::execute);
//    jdbcTemplate.execute("truncate finale_places", PreparedStatement::execute);
//    jdbcTemplate.execute("truncate game", PreparedStatement::execute);
//    jdbcTemplate.execute("truncate money_transfer", PreparedStatement::execute);
//    jdbcTemplate.execute("truncate person", PreparedStatement::execute);
//    jdbcTemplate.execute("truncate poll_task", PreparedStatement::execute);
//    jdbcTemplate.execute("truncate prize_pool", PreparedStatement::execute);
//    jdbcTemplate.execute("truncate withdrawal", PreparedStatement::execute);
  }


}

File: core/src/testFixtures/java/by/mrrockka/extension/CorePSQLExtension.kt
package by.mrrockka.extension

import by.mrrockka.extension.TestPSQLContainer.Companion.container
import by.mrrockka.extension.TestPSQLContainer.Companion.dbName
import by.mrrockka.extension.TestPSQLContainer.Companion.password
import by.mrrockka.extension.TestPSQLContainer.Companion.username
import by.mrrockka.extension.TestPSQLContainer.Companion.version
import by.mrrockka.repo.BountyTable
import by.mrrockka.repo.EntriesTable
import by.mrrockka.repo.FinalePlacesTable
import by.mrrockka.repo.GameTable
import by.mrrockka.repo.MoneyTransferTable
import by.mrrockka.repo.PersonTable
import by.mrrockka.repo.PrizePoolTable
import by.mrrockka.repo.WithdrawalTable
import org.jetbrains.exposed.sql.deleteAll
import org.junit.jupiter.api.extension.AfterEachCallback
import org.junit.jupiter.api.extension.BeforeAllCallback
import org.junit.jupiter.api.extension.ExtensionContext
import org.springframework.test.context.junit.jupiter.SpringExtension
import org.springframework.transaction.support.TransactionTemplate

open class CorePSQLExtension : BeforeAllCallback, AfterEachCallback {
    override fun beforeAll(context: ExtensionContext?) {
        container.start()
        System.setProperty("spring.datasource.url", "jdbc:tc:postgresql:$version:///%$dbName")
        System.setProperty("spring.datasource.username", username)
        System.setProperty("spring.datasource.password", password)
    }

    override fun afterEach(context: ExtensionContext) {
        val template = SpringExtension.getApplicationContext(context)
                .getBean(TransactionTemplate::class.java)
        template.setPropagationBehaviorName("PROPAGATION_REQUIRED")

        template.execute {
            cleanCoreTable()
        }
    }

    protected fun cleanCoreTable() {
        MoneyTransferTable.deleteAll()
        BountyTable.deleteAll()
        EntriesTable.deleteAll()
        WithdrawalTable.deleteAll()
        PrizePoolTable.deleteAll()
        FinalePlacesTable.deleteAll()
        PersonTable.deleteAll()
        GameTable.deleteAll()
    }

}
